### ISSUE TITLE
Separate Submit from Publish 

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxelib#142]
Allow submit without publishing, so people can test against a new version but will not automatically update unless it’s published.

We might want to keep the "submit" to auto-publish, and either have a flag on submit or another command (push?) then a new publish command.

**Comment by back2dos** [comment]
Seems that using `alpha` etc. actually solves this.

**Comment by ncannasse** [comment]
@back2dos are you sure ? does the server makes such exception when it comes to marking the current version ?

**Comment by back2dos** [comment]
I'm not sure I understand the ticket in the first place. Would not installing anything that is alpha/beta/rc unless explicitly specified solve this problem?

**Comment by jasononeil** [comment]
I'm happy to look into this.  It will require a change on the server API where it chooses the "current" version.  It needs to choose based on the semver, currently it chooses based on the date of submission (See https://github.com/HaxeFoundation/haxelib/blob/master/src/tools/haxelib/Repo.hx#L285).

There would be a few changes (improvements/fixes) to the semantics:

* The website lists would be sorted by semver, not upload time.
* The current version would be based on semver, not upload time, meaning:
     * You could upload a bug fix for a "2.*" version, when "3.*" is already uploaded, and "3.*" will remain current.
     * If there is a "stable" version (Version > 1.0, and not tagged alpha/beta/rc) then this should take precedence over a non-stable version.
* `haxelib install` should install the latest stable version (if there is a stable version).
* `haxelib update` should only update to the same level of release.  (If you have a stable release, only update to a stable release. If you have a beta though update to a beta).
* Perhaps we can have a `haxelib install project [alpha|beta|rc]` style commands.

In short, if we actually draw some meaning from our semantic versioning, I think we have an adequate solution to this problem.

**Comment by ncannasse** [comment]
I agree for everything apart from haxelib update sticking to the same level of release. I think this will cause much more issues that it fixes. Maybe we could instead have an extra optional parameter such as "haxelib update mylib 3.2" that would update to the latest 3.2.x

**Comment by back2dos** [comment]
You can always use `haxelib install <libname> <version>` to get a specific version. As it is, `haxelib update` is basically equivalent to `haxelib install <libname> latest`.

I also think what Jason proposes is to stick with the same level of a major release. Since major releases are meant to express a break in compatibility, that seems a sensible thing to do.

**Comment by back2dos** [comment]
For now I have implemented this by having the client filter for the latest stable, as `ProjectInfos` has the whole version log anyway. Does that solve the issue @ncannasse ? If so, please close.

Regarding all the other SemVer related stuff, I'm going to move this into another ticket, as that's out of scope of this issue.

**Comment by ncannasse** [comment]
@back2dos I'm not sure that it solves it or not, could you reference the corresponding commit ?

**Comment by back2dos** [comment]
This is basically it: https://github.com/HaxeFoundation/haxelib/commit/e8417a55d4f8bb18145389da325efb1713c96bc4

Although the current state is a little different now (sorting by semver version comparison): https://github.com/HaxeFoundation/haxelib/blob/master/src/tools/haxelib/Data.hx#L196

**Comment by ncannasse** [comment]
Looks fine to me. We still need to do a bit of SemVer sorting on the website, so assigning to @jasononeil

**Comment by jasononeil** [comment]
I'll track that in #176
